h/static/scripts/group-forms/hooks/use-update-moderation-status.ts

```diff
@@ -0,0 +1,78 @@
+import { useCallback, useContext, useEffect, useRef } from 'preact/hooks';
+
+import { Config } from '../config';
+import type { APIAnnotationData, ModerationStatus } from '../utils/api';
+import { callAPI } from '../utils/api';
+
+export type UpdateModerationStatusError = {
+  message: string;
+  error: any;
+};
+
+/**
+ * When invoked, it updates the annotation's moderation status via API call
+ *
+ * @param onError - A callback invoked if the API request fails
+ */
+export type UpdateModerationStatusCallback = (
+  newModerationStatus: ModerationStatus,
+  onError: (error: UpdateModerationStatusError) => void,
+) => Promise<void>;
+
+/**
+ * Returns a callback that can be used to update the moderation status of an
+ * annotation via API call
+ */
+export function useUpdateModerationStatus(
+  annotation: APIAnnotationData,
+): UpdateModerationStatusCallback {
+  const config = useContext(Config);
+
+  // Track the last in-flight request
+  const abortCtrlRef = useRef<AbortController>();
+  useEffect(() => {
+    const abortCtrl = abortCtrlRef.current;
+    return () => abortCtrl?.abort();
+  }, []);
+
+  return useCallback(
+    async (
+      moderationStatus: ModerationStatus,
+      onError: (error: UpdateModerationStatusError) => void,
+    ) => {
+      const annotationId = annotation.id;
+      if (!config?.api.annotationModeration || !annotationId) {
+        return;
+      }
+
+      const { url, ...apiConfig } = config.api.annotationModeration;
+
+      if (abortCtrlRef.current) {
+        abortCtrlRef.current.abort();
+      }
+      abortCtrlRef.current = new AbortController();
+
+      try {
+        await callAPI<APIAnnotationData>(
+          url.replace(':annotationId', annotationId),
+          {
+            ...apiConfig,
+            signal: abortCtrlRef.current.signal,
+            json: {
+              annotation_updated: annotation.updated,
+              moderation_status: moderationStatus,
+            },
+          },
+        );
+      } catch (error) {
+        // TODO Explicitly handle conflict errors
+        onError({
+          message: 'An error occurred updating the moderation status',
+          error,
+        });
+        throw error;
```

Comment:
As discussed in slack, we have decided to add an `onError` callback that enforces consumers to handle the errors produced while trying to change the moderation status.

However, while working on this I realized we need to run code conditionally in case of "no" error (specifically, the logic to update the annotation status locally, or remove it from the list).

In order to allow that, I rethrow the error here, so that we can still await on this action, and anything after won't be called if it rejects, but I think this might be confusion.

Perhaps the right approach would be to also have an `onSuccess` callback.

Reply:
Ok, looks like we both realized there was something odd here 😅 https://github.com/hypothesis/h/pull/9657#discussion_r2168984284